django/db/models/sql/compiler.py

```diff
@@ -894,6 +894,12 @@ def get_related_klass_infos(klass_info, related_klass_infos):
                     from_parent = issubclass(model, opts.model) and model is not opts.model
 
                     def local_setter(obj, from_obj):
+                        if from_obj is None:
+                            # In FilteredRelation with condition returns no value cases, form_obj is none
+                            # and FieldCacheMixin.set_cached_value raises an error about NoneType object.
+                            # form_obj has to be derived from a related class of the join as a key of caching.
+                            if self.klass_info['related_klass_infos']:
+                                from_obj = self.klass_info['related_klass_infos'][0]['model']()
```

Comment:
Creating a new empty instance of related `model` doesn't look like a proper solution for me. IMO we should just omit `set_cached_value()` when `from_obj` is not available e.g.
```python
def local_setter(obj, from_obj):
    if from_obj:
        f.remote_field.set_cached_value(from_obj, obj)
```

Reply:
in this case, `from_obj` won't be set with None value. this is a problem, isn't it?

Reply:
This is reverse relation so your solution will set `None` in a new empty and not really existing instance. `set_cached_value()` should set object for a reverse relation so there is not reason to set it for a mocked empty instance which doesn't exist outside this method. For example, you would basically do this in added test:
```python
Book()._state.fields_cache['author'] = <Author: Alice>
Book()._state.fields_cache['author'] = <Author: Jane>
```

Reply:
Thanks for the explation.